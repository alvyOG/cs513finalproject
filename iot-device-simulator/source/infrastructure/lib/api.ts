
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, Construct, RemovalPolicy } from '@aws-cdk/core';
import {
  AccessLogFormat,
  AuthorizationType,
  ContentHandling,
  Deployment,
  EndpointType,
  Integration,
  IntegrationType,
  LogGroupLogDestination,
  MethodLoggingLevel,
  MethodOptions,
  PassthroughBehavior,
  RequestValidator,
  RestApi,
  Stage
} from '@aws-cdk/aws-apigateway';
import { Policy, ServicePrincipal } from '@aws-cdk/aws-iam';
import { LogGroup, RetentionDays } from '@aws-cdk/aws-logs';
import { addCfnSuppressRules } from '../utils/utils';
import { Function as LambdaFunction } from '@aws-cdk/aws-lambda';
import { IBucket } from '@aws-cdk/aws-s3';
import { Table } from '@aws-cdk/aws-dynamodb';

/**
 * ApiConstructProps props
 * @interface ApiConstructProps
 */
export interface ApiConstructProps {
  // Policy for CloudWatch Logs
  readonly cloudWatchLogsPolicy: Policy;
  // IoT endpoint address
  readonly stepFunctionsARN: string;
  // Simulation data DynamoDB table name
  readonly simulationTable: Table;
  // Device Type data DynamoDB table name
  readonly deviceTypeTable: Table;
  //Routes S3 bucket
  readonly routesBucketArn: string;
  //microservices lambda
  readonly microservicesLambda: LambdaFunction;
  /**
   * Solution config properties.
   * Logging level, solution ID, version, source code bucket, and source code prefix
   */
  readonly solutionConfig: {
    sendAnonymousUsage: string;
    solutionId: string;
    solutionVersion: string;
    sourceCodeBucket: IBucket;
    sourceCodePrefix: string;
  };
  // Solution UUID
  readonly uuid: string;
}

/**
 * @class
 * IoT Device Simulator Framework API Construct.
 * It creates an API Gateway REST API and other resources.
 */
export class ApiConstruct extends Construct {
  // API endpoint
  public apiEndpoint: string;
  // API ID
  public apiId: string;
  //microservices lambda
  public microservicesLambdaFunction: LambdaFunction;

  constructor(scope: Construct, id: string, props: ApiConstructProps) {
    super(scope, id);

    const apiLogGroup = new LogGroup(this, 'Logs', {
      removalPolicy: RemovalPolicy.DESTROY,
      retention: RetentionDays.THREE_MONTHS
    });
    addCfnSuppressRules(apiLogGroup, [{
      id: 'W84', reason: 'CloudWatch Logs are already encrypted by default.'
    }]);

    const api = new RestApi(this, 'IoTDeviceSimulatorApi', {
      defaultCorsPreflightOptions: {
        allowOrigins: ['*'],
        allowHeaders: [
          'Authorization',
          'Content-Type',
          'X-Amz-Date',
          'X-Amz-Security-Token',
          'X-Api-Key'
        ],
        allowMethods: [
          'GET',
          'POST',
          'PUT',
          'DELETE',
          'OPTIONS'
        ],
        statusCode: 200
      },
      deploy: true,
      deployOptions: {
        accessLogDestination: new LogGroupLogDestination(apiLogGroup),
        accessLogFormat: AccessLogFormat.jsonWithStandardFields(),
        loggingLevel: MethodLoggingLevel.INFO,
        stageName: 'prod',
        tracingEnabled: true
      },
      description: 'IoT Device Simulator Rest API',
      endpointTypes: [EndpointType.REGIONAL]
    });
    this.apiEndpoint = `https://${api.restApiId}.execute-api.${Aws.REGION}.amazonaws.com/prod`;
    this.apiId = api.restApiId;

    const requestValidator = new RequestValidator(this, 'ApiRequestValidator', {
      restApi: api,
      validateRequestParameters: true,
      validateRequestBody: true
    });

    addCfnSuppressRules(api.node.findChild('Deployment') as Deployment, [{
      id: 'W68', reason: 'The solution does not require the usage plan.'
    }]);
    addCfnSuppressRules(api.node.findChild('DeploymentStage.prod') as Stage, [{
      id: 'W64', reason: 'The solution does not require the usage plan.'
    }]);

    /**
     * method options for all methods
     */
    const universalMethodOptions: MethodOptions = {
      authorizationType: AuthorizationType.IAM,
      methodResponses: [{
        statusCode: '200',
        responseModels: {
          'application/json': { modelId: 'Empty' }
        }
      }],
      requestParameters: { 'method.request.querystring.nextToken': false },
      requestValidator: requestValidator
    };

    /**
     * Integration for all resources
     */
    const universalIntegration = new Integration({
      type: IntegrationType.AWS_PROXY,
      integrationHttpMethod: 'POST',
      options: {
        contentHandling: ContentHandling.CONVERT_TO_TEXT,
        integrationResponses: [{ statusCode: '200' }],
        passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH
      },
      uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${props.microservicesLambda.functionArn}/invocations`
    });

    /**
     * simulation API
     * ANY /simulation
     * ANY /simulation/{simid}
     */
    const simulationResource = api.root.addResource('simulation');
    simulationResource.addMethod(
      'ANY',
      universalIntegration,
      universalMethodOptions
    );
    const simulationSimIdResource = simulationResource.addResource('{simid}');
    simulationSimIdResource.addMethod('ANY', universalIntegration, universalMethodOptions);

    /**
     * Devive Types API
     * ANY /devicetypes
     * ANY /devicetypes/{typeid}
     */
    const deviceTypesResource = api.root.addResource('devicetypes');
    deviceTypesResource.addMethod('ANY', universalIntegration, universalMethodOptions);
    const typeIdResource = deviceTypesResource.addResource('{typeid}');
    typeIdResource.addMethod('ANY', universalIntegration, universalMethodOptions);

    props.microservicesLambda.addPermission('ApiLambdaInvokePermission', {
      action: 'lambda:InvokeFunction',
      principal: new ServicePrincipal('apigateway.amazonaws.com'),
      sourceArn: api.arnForExecuteApi()
    });
  }
}