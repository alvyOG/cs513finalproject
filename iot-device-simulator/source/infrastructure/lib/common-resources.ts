// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ArnFormat, Construct, RemovalPolicy, Stack } from '@aws-cdk/core';
import { Effect, Policy, PolicyStatement } from '@aws-cdk/aws-iam';
import { BlockPublicAccess, Bucket, BucketAccessControl, BucketEncryption, IBucket } from '@aws-cdk/aws-s3';
import { addCfnSuppressRules } from '../utils/utils';

/**
 * CommonResourcesConstruct props
 * @interface CommonResourcesConstructProps
 */
export interface CommonResourcesConstructProps {
  // Source code bucket
  sourceCodeBucket: string;
}

/**
 * @class
 * IoT Device Simulator Common Resources Construct.
 * It creates a common CloudWatch Logs policy for Lambda functions, a logging S3 bucket, and M2C2 S3 bucket.
 */
export class CommonResourcesConstruct extends Construct {
  // CloudWatch Logs policy
  public cloudWatchLogsPolicy: Policy;
  // S3 Logging bucket
  public s3LoggingBucket: Bucket;
  // Code S3 bucket
  public sourceCodeBucket: IBucket;

  constructor(scope: Construct, id: string, props: CommonResourcesConstructProps) {
    super(scope, id);

    this.cloudWatchLogsPolicy = new Policy(this, 'CloudWatchLogsPolicy', {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:PutLogEvents'
          ],
          resources: [
            Stack.of(this).formatArn({ service: 'logs', resource: 'log-group', resourceName: '/aws/lambda/*', arnFormat: ArnFormat.COLON_RESOURCE_NAME })
          ]
        })
      ]
    });

    this.s3LoggingBucket = new Bucket(this, 'LogBucket', {
      accessControl: BucketAccessControl.LOG_DELIVERY_WRITE,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
      removalPolicy: RemovalPolicy.RETAIN
    });

    addCfnSuppressRules(this.s3LoggingBucket, [
      { id: 'W35', reason: 'This bucket is to store S3 logs, so it does not require access logs.' },
      { id: 'W51', reason: 'This bucket is to store S3 logs, so it does not require S3 policy.' }
    ]);

    this.sourceCodeBucket = Bucket.fromBucketName(this, 'SourceCodeBucket', props.sourceCodeBucket);
  }
}