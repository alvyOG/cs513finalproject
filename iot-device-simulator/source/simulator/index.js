// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

const Engine = require('./lib/engine');
const AWS = require('aws-sdk');
const iot = new AWS.Iot({ region: process.env.AWS_REGION });

exports.handler = async (event, context) => {

    'use strict';
    context.callbackWaitsForEmptyEventLoop = false;
    let results = [];
    let options = event.options ? event.options : {};
    let simulation = event.simulation;
    let devices = event.devices ? event.devices : simulation.devices;
    delete simulation.devices;
    options.context = context;

    //Restart already started simulation
    if (options.restart === true) {
        options.restart = false;
        results = await Promise.all(devices.map(async (device) => {
            try {
                const engine = new Engine(options, simulation, device);
                return engine.start();
            } catch (err) {
                console.error(err);
                throw err;
            }
        }))
    }
    else {
        //start engine for each device type
        results = await Promise.all(
            devices.map(async device => {
                device.info = AWS.DynamoDB.Converter.unmarshall(device.info);
                try {
                    // create engine instance and start
                    const engine = new Engine(options, simulation, device);
                    return engine.start();
                } catch (err) {
                    throw err;
                }
            }));
    }
    //make sure every device for every device type is completed
    options.restart = !results.every((result) => result === 'complete');
    //delete unncessary information to reduce data passed through step fucntions
    delete options.context;
    delete options.currentState;
    delete options.staticValues;
    results = {
        options: options,
        simulation: simulation,
        devices: results
    }
    return results;
};


