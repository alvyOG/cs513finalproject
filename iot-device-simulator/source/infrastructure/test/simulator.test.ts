// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { Bucket } from '@aws-cdk/aws-s3';
import { Policy, PolicyStatement } from '@aws-cdk/aws-iam';
import { SimulatorConstruct } from '../lib/simulator';
import { AttributeType, Table } from '@aws-cdk/aws-dynamodb';


test('IoT Device Simulator SimulatorConstruct Test', () => {
    const stack = new Stack();
    const testRouteBucket = new Bucket(stack, 'testRouteBucket');
    const testPolicy = new Policy(stack, 'TestPolicy', {
        statements: [
            new PolicyStatement({
                resources: ['*'],
                actions: ['cloudwatch:Get*']
            })
        ]
    });
    const testDTypeTable = new Table(stack, 'TestDTypeTable', {
        partitionKey: {
            name: "typeId", type: AttributeType.STRING
        }
    });
    const testSimTable = new Table(stack, 'TestSimTable', {
        partitionKey: {
            name: "simId", type: AttributeType.STRING
        }
    });
    const testSourceBucket = Bucket.fromBucketName(stack, 'SourceCodeBucket', 'test-bucket-region');

    const simulator = new SimulatorConstruct(stack, 'simulator', {
        cloudWatchLogsPolicy: testPolicy,
        iotEndpointAddress: 'abcd123efg45h-ats.iot.some-region.amazonaws.com',
        simulationTable: testSimTable,
        deviceTypeTable: testDTypeTable,
        routesBucket: testRouteBucket,
        uniqueSuffix: "testSuffix",
        solutionConfig: {
            sendAnonymousUsage: 'Yes',
            solutionId: 'testId',
            solutionVersion: 'testVersion',
            sourceCodeBucket: testSourceBucket,
            sourceCodePrefix: 'testPrefix/',
        },
        uuid: 'abc123'
    })

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});