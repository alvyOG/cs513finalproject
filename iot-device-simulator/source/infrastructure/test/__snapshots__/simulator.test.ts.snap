// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IoT Device Simulator SimulatorConstruct Test 1`] = `
Object {
  "Resources": Object {
    "TestDTypeTableEB82FCB5": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "typeId",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "typeId",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "TestPolicyCC05E598": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudwatch:Get*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestPolicyCC05E598",
        "Roles": Array [
          Object {
            "Ref": "simulatorEngineLambdaRoleE774CD29",
          },
          Object {
            "Ref": "simulatorMicroservicesRole0AC1A17A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestSimTable7336F78A": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "simId",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "simId",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "simulatorEngineLambda17530EFE": Object {
      "DependsOn": Array [
        "simulatorEngineLambdaRoleDefaultPolicy9ECE33DE",
        "simulatorEngineLambdaRoleE774CD29",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "test-bucket-region",
          "S3Key": "testPrefix//simulator.zip",
        },
        "Description": "IoT Device Simulator function",
        "Environment": Object {
          "Variables": Object {
            "IOT_ENDPOINT": "abcd123efg45h-ats.iot.some-region.amazonaws.com",
            "ROUTE_BUCKET": Object {
              "Ref": "testRouteBucket64D9BEB5",
            },
            "SEND_ANONYMOUS_METRIC": "Yes",
            "SIM_TABLE": Object {
              "Ref": "TestSimTable7336F78A",
            },
            "SOLUTION_ID": "testId",
            "UUID": "abc123",
            "VERSION": "testVersion",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "simulatorEngineLambdaRoleE774CD29",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "simulatorEngineLambdaRoleDefaultPolicy9ECE33DE": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testRouteBucket64D9BEB5",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testRouteBucket64D9BEB5",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "simulatorEngineLambdaRoleDefaultPolicy9ECE33DE",
        "Roles": Array [
          Object {
            "Ref": "simulatorEngineLambdaRoleE774CD29",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "simulatorEngineLambdaRoleE774CD29": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Path": "/",
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "s3:GetObject",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "testRouteBucket64D9BEB5",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "S3Policy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "dynamodb:GetItem",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "TestSimTable7336F78A",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "DynamoDBPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "iot:Publish",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":iot:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":topic/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "IoTPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "simulatorMicroservicesRole0AC1A17A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Path": "/",
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "dynamodb:PutItem",
                    "dynamodb:DeleteItem",
                    "dynamodb:GetItem",
                    "dynamodb:Scan",
                    "dynamodb:Query",
                    "dynamodb:BatchGetItem",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "TestSimTable7336F78A",
                        "Arn",
                      ],
                    },
                    Object {
                      "Fn::GetAtt": Array [
                        "TestDTypeTableEB82FCB5",
                        "Arn",
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "DynamoDBPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "simulatorMicroservicesRoleDefaultPolicy8F21950B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "simulatorStepFunctionsStateMachineA2150DAB",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "simulatorMicroservicesRoleDefaultPolicy8F21950B",
        "Roles": Array [
          Object {
            "Ref": "simulatorMicroservicesRole0AC1A17A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "simulatorStepFunctionsExecutionAbortedAlarmC24E8F0E": Object {
      "Properties": Object {
        "AlarmDescription": "Alarm for the number of executions that aborted exceeded the threshold of 1. ",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StateMachineArn",
            "Value": Object {
              "Ref": "simulatorStepFunctionsStateMachineA2150DAB",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ExecutionsAborted",
        "Namespace": "AWS/States",
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "simulatorStepFunctionsExecutionFailedAlarmACE1FB5D": Object {
      "Properties": Object {
        "AlarmDescription": "Alarm for the number of executions that failed exceeded the threshold of 1. ",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StateMachineArn",
            "Value": Object {
              "Ref": "simulatorStepFunctionsStateMachineA2150DAB",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ExecutionsFailed",
        "Namespace": "AWS/States",
        "Period": 300,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "simulatorStepFunctionsExecutionThrottledAlarm542D795D": Object {
      "Properties": Object {
        "AlarmDescription": "Alarm for the number of executions that throttled exceeded the threshold of 1. ",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StateMachineArn",
            "Value": Object {
              "Ref": "simulatorStepFunctionsStateMachineA2150DAB",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ExecutionThrottled",
        "Namespace": "AWS/States",
        "Period": 300,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "simulatorStepFunctionsLogGroupF673A0C9": Object {
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W84",
              "reason": "KMS encryption unnecessary for log group",
            },
          ],
        },
      },
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/vendedlogs/states/",
              Object {
                "Ref": "AWS::StackName",
              },
              "-simulatorStepFunctionsLogGroup-testSuffix",
            ],
          ],
        },
        "RetentionInDays": 365,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "simulatorStepFunctionsStateMachineA2150DAB": Object {
      "DependsOn": Array [
        "simulatorStepFunctionsStateMachineRoleDefaultPolicy02630DD3",
        "simulatorStepFunctionsStateMachineRole6FCF7FE0",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W11",
              "reason": "CloudWatch logs actions do not support resource level permissions",
            },
            Object {
              "id": "W12",
              "reason": "CloudWatch logs actions do not support resource level permissions",
            },
          ],
        },
      },
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"getDeviceTypeMap\\",\\"States\\":{\\"getDeviceTypeMap\\":{\\"Type\\":\\"Map\\",\\"ResultPath\\":\\"$.simulation.devices\\",\\"Next\\":\\"simulatorInvoke\\",\\"Parameters\\":{\\"typeId.$\\":\\"$$.Map.Item.Value.typeId\\",\\"amount.$\\":\\"$$.Map.Item.Value.amount\\"},\\"Iterator\\":{\\"StartAt\\":\\"getDeviceTypeInfo\\",\\"States\\":{\\"getDeviceTypeInfo\\":{\\"End\\":true,\\"Type\\":\\"Task\\",\\"ResultPath\\":\\"$.info\\",\\"ResultSelector\\":{\\"name.$\\":\\"$.Item.name\\",\\"topic.$\\":\\"$.Item.topic\\",\\"payload.$\\":\\"$.Item.payload\\"},\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::dynamodb:getItem\\",\\"Parameters\\":{\\"Key\\":{\\"typeId\\":{\\"S.$\\":\\"$.typeId\\"}},\\"TableName\\":\\"",
              Object {
                "Ref": "TestDTypeTableEB82FCB5",
              },
              "\\",\\"ConsistentRead\\":false}}}},\\"ItemsPath\\":\\"$.simulation.devices\\",\\"MaxConcurrency\\":0},\\"simulatorInvoke\\":{\\"Next\\":\\"devicesRunning?\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Catch\\":[{\\"ErrorEquals\\":[\\"States.ALL\\"],\\"ResultPath\\":\\"$.error\\",\\"Next\\":\\"UpdateSimTable\\"}],\\"Type\\":\\"Task\\",\\"OutputPath\\":\\"$.Payload\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "simulatorEngineLambda17530EFE",
                  "Arn",
                ],
              },
              "\\",\\"Payload.$\\":\\"$\\"}},\\"devicesRunning?\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.options.restart\\",\\"BooleanEquals\\":true,\\"Next\\":\\"simulatorInvoke\\"}],\\"Default\\":\\"UpdateSimTable\\"},\\"UpdateSimTable\\":{\\"Next\\":\\"Done\\",\\"Catch\\":[{\\"ErrorEquals\\":[\\"DynamoDB.ConditionalCheckFailedException\\"],\\"Next\\":\\"Done\\"}],\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::dynamodb:updateItem\\",\\"Parameters\\":{\\"Key\\":{\\"simId\\":{\\"S.$\\":\\"$.simulation.simId\\"}},\\"TableName\\":\\"",
              Object {
                "Ref": "TestSimTable7336F78A",
              },
              "\\",\\"ConditionExpression\\":\\"attribute_exists(simId)\\",\\"ExpressionAttributeValues\\":{\\":stage\\":{\\"S\\":\\"sleeping\\"},\\":time\\":{\\"S.$\\":\\"$$.State.EnteredTime\\"}},\\"UpdateExpression\\":\\"SET stage = :stage, updatedAt = :time\\"}},\\"Done\\":{\\"Type\\":\\"Succeed\\"}}}",
            ],
          ],
        },
        "LoggingConfiguration": Object {
          "Destinations": Array [
            Object {
              "CloudWatchLogsLogGroup": Object {
                "LogGroupArn": Object {
                  "Fn::GetAtt": Array [
                    "simulatorStepFunctionsLogGroupF673A0C9",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "IncludeExecutionData": false,
          "Level": "ALL",
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "simulatorStepFunctionsStateMachineRole6FCF7FE0",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "simulatorStepFunctionsStateMachineRole6FCF7FE0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "simulatorStepFunctionsStateMachineRoleDefaultPolicy02630DD3": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "The 'LogDelivery' actions do not support resource-level authorizations",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogDelivery",
                "logs:GetLogDelivery",
                "logs:UpdateLogDelivery",
                "logs:DeleteLogDelivery",
                "logs:ListLogDeliveries",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "simulatorEngineLambda17530EFE",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "dynamodb:UpdateItem",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":dynamodb:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":table/",
                    Object {
                      "Ref": "TestSimTable7336F78A",
                    },
                  ],
                ],
              },
            },
            Object {
              "Action": "dynamodb:GetItem",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":dynamodb:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":table/",
                    Object {
                      "Ref": "TestDTypeTableEB82FCB5",
                    },
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "logs:PutResourcePolicy",
                "logs:DescribeResourcePolicies",
                "logs:DescribeLogGroups",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "simulatorStepFunctionsStateMachineRoleDefaultPolicy02630DD3",
        "Roles": Array [
          Object {
            "Ref": "simulatorStepFunctionsStateMachineRole6FCF7FE0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "simulatormicroservices752C24A5": Object {
      "DependsOn": Array [
        "simulatorMicroservicesRoleDefaultPolicy8F21950B",
        "simulatorMicroservicesRole0AC1A17A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "test-bucket-region",
          "S3Key": "testPrefix//microservices.zip",
        },
        "Description": "IoT Device Simulator microservices function",
        "Environment": Object {
          "Variables": Object {
            "DEVICE_TYPES_TBL": Object {
              "Ref": "TestDTypeTableEB82FCB5",
            },
            "SEND_ANONYMOUS_METRIC": "Yes",
            "SIMULATIONS_TBL": Object {
              "Ref": "TestSimTable7336F78A",
            },
            "SIM_STEP_FUNCTION": Object {
              "Ref": "simulatorStepFunctionsStateMachineA2150DAB",
            },
            "SOLUTION_ID": "testId",
            "UUID": "abc123",
            "VERSION": "testVersion",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "simulatorMicroservicesRole0AC1A17A",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "testRouteBucket64D9BEB5": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
